onset_maproom/calc.py

```diff
@@ -53,6 +332,228 @@ def water_balance(
+    # Create or Initialize Kc
+    if kc_params is None:
+        kc0 = xr.ones_like(water_balance.et)
+    else:
+        if p_d is not None:
+            kc0 = kc_interpolation(p_d, kc_params, time_coord=time_coord)
+        else:
+            p_d_find = planting_date(
+                sminit0, sm_threshold, time_coord=time_coord
+            ).expand_dims(dim=time_coord)
+            kc0 = kc_interpolation(p_d_find, kc_params, time_coord=time_coord)
+    # Initializaing sm
+    if time_coord in kc0:
+        kc0 = kc0.sel({time_coord: water_balance.soil_moisture[time_coord][0]})
+    water_balance.et_crop[{time_coord: 0}] = crop_evapotranspiration(
+        water_balance.et.isel({time_coord: 0}).expand_dims(dim=time_coord),
+        kc0,
+        time_coord=time_coord,
+    ).squeeze(time_coord)
+    water_balance.soil_moisture[{time_coord: 0}] = (
+        sminit
+        + water_balance.peffective.isel({time_coord: 0})
+        - reduce_crop_evapotranspiration(
+            water_balance.et_crop
+            .isel({time_coord: 0}).expand_dims(dim=time_coord),
+            ks,
+        ).squeeze(time_coord)
+    ).clip(0, taw)
+    # Looping on time_coord
+    for i in range(1, time_coord_size):
+        if rho is not None:
+            ks = single_stress_coeff(
+                water_balance.soil_moisture
+                .isel({time_coord: i - 1}).expand_dims(dim=time_coord),
+                taw,
+                raw,
+                time_coord=time_coord,
+            ).squeeze(time_coord)
+        if kc_params is not None and p_d is None:
+            p_d_find = planting_date(
+                xr.concat(
+                    [
+                        sminit0,
+                        water_balance.soil_moisture
+                        .isel({time_coord: slice(0, i - 1)}),
+                    ],
+                    time_coord,
+                ),
+                sm_threshold,
+                time_coord=time_coord,
+            ).expand_dims(dim=time_coord)
+            kc = kc_interpolation(p_d_find, kc_params, time_coord=time_coord)
+        else:
+            kc = kc0
+        if time_coord in kc:
+            kc = kc.sel({time_coord: water_balance.soil_moisture[time_coord][i]})
+        water_balance.et_crop[{time_coord: i}] = crop_evapotranspiration(
+            water_balance.et
+            .isel({time_coord: i}).expand_dims(dim=time_coord),
+            kc,
+            time_coord=time_coord,
+        ).squeeze(time_coord)
+        water_balance.soil_moisture[{time_coord: i}] = (
+            water_balance.soil_moisture.isel({time_coord: i - 1}, drop=True)
+            + water_balance.peffective.isel({time_coord: i}, drop=True)
+            - reduce_crop_evapotranspiration(
+                water_balance.et_crop
+                .isel({time_coord: i}).expand_dims(dim=time_coord),
+                ks,
+            ).squeeze(time_coord)
+        ).clip(0, taw)
+    # Save planting date if computed
```

Comment:
I would suggest that it's better to save it regardless of whether it was computed or provided as input. Being inconsistent about which variables are included could lead to errors when people assume that something they see in an example output will always be present.

Reply:
Can do. There is however the case where kc_params = None where there is no planting date and Kc is always 1. Should I still save p_d as null or something?

Reply:
The two options that come to mind are to omit it in that case, or to make it an array of NaNs with the appropriate shape. Given my ignorance of how it will be used, I don't know which would be better. Probably not a very consequential decision.